Implement image reader

rizwangilani opened the pull request:
Current implementation supports jpeg, png, and mxnet recordio file formats

*Testing*:

```
ctest --verbose
...
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 1
    Start 1: mlio-test
...

1: [==========] Running 19 tests from 4 test cases.
1: [----------] Global test environment set-up.
1: [----------] 2 tests from test_blob_record_reader
1: [ RUN      ] test_blob_record_reader.test_read_record
1: [       OK ] test_blob_record_reader.test_read_record (0 ms)
1: [ RUN      ] test_blob_record_reader.test_peek_record
1: [       OK ] test_blob_record_reader.test_peek_record (0 ms)
1: [----------] 2 tests from test_blob_record_reader (0 ms total)
1:
1: [----------] 3 tests from test_text_line_reader
1: [ RUN      ] test_text_line_reader.test_text_line_reader_happy_path
1: [       OK ] test_text_line_reader.test_text_line_reader_happy_path (2 ms)
1: [ RUN      ] test_text_line_reader.test_text_line_reader_batch_greater_than_features
1: [       OK ] test_text_line_reader.test_text_line_reader_batch_greater_than_features (1 ms)
1: [ RUN      ] test_text_line_reader.test_text_line_reader_batch_less_than_features
1: [       OK ] test_text_line_reader.test_text_line_reader_batch_less_than_features (0 ms)
1: [----------] 3 tests from test_text_line_reader (3 ms total)
1:
1: [----------] 3 tests from test_recordio_protobuf_reader
1: [ RUN      ] test_recordio_protobuf_reader.test_complete_records_path
1: [       OK ] test_recordio_protobuf_reader.test_complete_records_path (0 ms)
1: [ RUN      ] test_recordio_protobuf_reader.test_split_records_path
1: [       OK ] test_recordio_protobuf_reader.test_split_records_path (1 ms)
1: [ RUN      ] test_recordio_protobuf_reader.test_corrupt_split_records_patH
1: [       OK ] test_recordio_protobuf_reader.test_corrupt_split_records_patH (1 ms)
1: [----------] 3 tests from test_recordio_protobuf_reader (2 ms total)
1:
1: [----------] 11 tests from test_image_reader
1: [ RUN      ] test_image_reader.test_jpg_3_channel_last_batch_dropped
1: [       OK ] test_image_reader.test_jpg_3_channel_last_batch_dropped (6 ms)
1: [ RUN      ] test_image_reader.test_jpg_3_channel_with_resize_last_batch_dropped
1: [       OK ] test_image_reader.test_jpg_3_channel_with_resize_last_batch_dropped (7 ms)
1: [ RUN      ] test_image_reader.test_jpg_1_channel
1: [       OK ] test_image_reader.test_jpg_1_channel (2 ms)
1: [ RUN      ] test_image_reader.test_png_3_channel_happy_path
1: [       OK ] test_image_reader.test_png_3_channel_happy_path (2 ms)
1: [ RUN      ] test_image_reader.test_png_3_channel_happy_path_with_resize
1: [       OK ] test_image_reader.test_png_3_channel_happy_path_with_resize (1 ms)
1: [ RUN      ] test_image_reader.test_png_1_channel
1: [       OK ] test_image_reader.test_png_1_channel (2 ms)
1: [ RUN      ] test_image_reader.test_invalid_image_dimensions_parameter
1: [       OK ] test_image_reader.test_invalid_image_dimensions_parameter (11 ms)
1: [ RUN      ] test_image_reader.test_bgr_to_rgb
1: [       OK ] test_image_reader.test_bgr_to_rgb (2 ms)
1: [ RUN      ] test_image_reader.test_recordio_3_channel
1: [       OK ] test_image_reader.test_recordio_3_channel (1 ms)
1: [ RUN      ] test_image_reader.test_recordio_1_channel
1: [       OK ] test_image_reader.test_recordio_1_channel (1 ms)
1: [ RUN      ] test_image_reader.test_bad_batch_handling_skip
1: [       OK ] test_image_reader.test_bad_batch_handling_skip (5 ms)
1: [----------] 11 tests from test_image_reader (40 ms total)
1:
1: [----------] Global test environment tear-down
1: [==========] 19 tests from 4 test cases ran. (46 ms total)
1: [  PASSED  ] 19 tests.
1/1 Test #1: mlio-test ........................   Passed    0.07 sec

100% tests passed, 0 tests failed out of 1
```

Commit:
Implement image reader

Current implementation supports jpeg, png, and mxnet recordio file formats.

aws-patlin commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,224 @@
+        params_.img_frame != image_frame::pipe) {
+        throw std::invalid_argument{"Please enter a valid value for "
+                                    "image_frame parameter."};
+    }
```
Does it make sense to check parameters here instead of in the constructor?

cbalioglu replied:
> Does it make sense to check parameters here instead of in the constructor?

+1

cbalioglu replied:
And as a side note; if you wanted to keep the validation logic here, simply throw an error after the `switch` statement.

aws-patlin commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,224 @@
+        params_.data_shape[0] * params_.data_shape[1] * params_.data_shape[2];
+    auto total_memory_buffer =
+        static_cast<unsigned long>(instance_offset) * batch_size;
+    std::vector<std::uint8_t> img_vec(total_memory_buffer);
```
You should probably use list initialization here instead: `img_vec{total_memory_buffer}`

cbalioglu replied:
> You should probably use list initialization here instead: `img_vec{total_memory_buffer}`

He is allocating `img_vec` with a size of `total_memory_buffer` here. Using list initialization would cause `img_vec` to contain a single element with the value of `total_memory_buffer`. This is one of the areas where `initializer_list`s should not be used.

aws-patlin commented on include/mlio/image_reader.h:
```diff
@@ -0,0 +1,97 @@
+
+struct MLIO_API image_reader_params final {
+    /// Enum to identify 'file' or 'pipe' mode.
+    image_frame img_frame{};
```
Are we breaking any abstraction barriers by requiring the image reader to know the input mode of the data? How does this compare to other data readers?

rizwangilani replied:
hmm I don't think so. Do you suggest not having the image_frame param? Also we will default to `file` for python API.

As for the comparison, this is similar to how we have csv params: https://github.com/awslabs/ml-io/blob/master/include/mlio/csv_reader.h#L49

aws-patlin replied:
It doesn't seem like this is necessary for any other data reader though. Why is it that in this case, a difference in data source changes how the data reader needs to behave? Specifically, [this](https://github.com/rizwangilani/ml-io/blob/imreader/src/mlio/image_reader.cxx#L108-L113) is the only place where `img_frame` is used, and I'm wondering if there's a way to handle this before it gets to the image reader.

rizwangilani replied:
we are also using `img_frame` to identify what type of `record_reader` to we want to build in `make_record_reader()`. Alternatively, we can create 2 different image readers. however, that might be an overkill at this point. thoughts?

aws-patlin commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,224 @@
+    }
+
+    auto ptr = wrap_cpu_array<data_type::uint8>(std::move(img_vec));
+    auto tsr = make_intrusive<dense_tensor>(
```
Is it pretty standard across ML-IO to call the returned tensor `tsr`?

rizwangilani replied:
`tsr` seems to be used to refer to a tensor at various places. However, we can refactor this if this sounds confusing to you

aws-patlin replied:
No need, just wanted to clarify this was the case.

aws-patlin commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,224 @@
+{
+    cv::Mat decoded_img = cv::imdecode(buf, mode);
+    if (decoded_img.empty()) {
+        throw std::runtime_error{fmt::format("image decoding failed. Image: ",
```
Should this be `"Image decoding failed. Image: {0:s}"`?

rizwangilani replied:
yes. good catch. thanks

aws-patlin commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,224 @@
+/*
```
Can you add comments to make it easier to understand what each function is doing?

rizwangilani replied:
I have tried to name each function in a way that makes its purpose clear. for example `decode_image` is decoding the image and `crop` is cropping the region of interest. Do you want more detailed comments regarding the logic within functions?

https://refactoring.guru/smells/comments
`The best comment is a good name for a method or class.`

aws-patlin replied:
Yeah, I think in this case it would be helpful to have a few more detailed comments in the longer functions. Otherwise it's hard for a reader to understand conceptually how each function is being accomplished.

aws-patlin commented on tests/mlio-test/test_image_reader.cxx:
```diff
@@ -0,0 +1,439 @@
+#include <gtest/gtest.h>
```
I think some of these tests would benefit from having comments added to clarify what is being tested. It would also be helpful in the future if someone uses these tests as reference for using the C++ API.

rizwangilani replied:
I have named each test in a way that makes its purpose clear, and have tried to follow the `Arrange Act Assert` pattern. What type of comments are you referring to here? Do you mean comments within the tests? Just making sure that my understanding is correct :)

aws-patlin replied:
Yup, just comments within the test. If most of the tests are similar, maybe you only need to do so for a couple of them.

aws-patlin reviewed (commented):
A few nit-picks and suggestions.

cbalioglu commented on include/mlio/image_reader.h:
```diff
@@ -0,0 +1,98 @@
+
+#pragma once
+
+#include <opencv4/opencv2/core/mat.hpp>
```
You should forward declare `cv::Mat` here instead of including the header file. This forces all library users to have OpenCV installed even if they don't intend to use the image reader as this file is included in `mlio.h`.

```
namespace cv {
  class Mat;
}  // namespace cv
```

cbalioglu commented on include/mlio/image_reader.h:
```diff
@@ -0,0 +1,98 @@
+
+#pragma once
+
+#include <opencv4/opencv2/core/mat.hpp>
```
When I tried to compile ML-IO with this line, it failed. At least in Conda the correct path is `<opencv2/core/mat.hpp>`. `opencv4` is not included.

rizwangilani replied:
It is because the library has not been added using cmake. It will build once this is approved and rebased: https://github.com/rizwangilani/ml-io/pull/2

cbalioglu replied:
I manually added the library, but it still fails, because you are explicitly including opencv header files in `image_reader.h`.

rizwangilani replied:
got it, thanks!

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,224 @@
+                           image_reader_params image_prm)
+    : parallel_data_reader{std::move(rdr_prm)}, params_{std::move(image_prm)}
+{
+    if (params_.data_shape.size() != data_shape_size_) {
```
Ideally we should be able to infer the dimensions from the images themselves. Let's keep this as it is, but this needs to be definitely improved in the near future.

rizwangilani replied:
This logic is here as we want to make images uniform in terms of size. Input images could be of variable size (as mentioned by our NYC team customers). This is again inspired from mxnet. data_shape param ensures that every image in the tensor is of the same size. What do you think?

cbalioglu replied:
I am okay with it for the time-being, but in the future I like us to offer a more flexible API.

cbalioglu commented on include/mlio/image_reader.h:
```diff
@@ -0,0 +1,98 @@
+/// @addtogroup data_readers Data Readers
+/// @{
+
+enum class image_frame { file, pipe };
```
`pipe` is not a right name here. We are naming the "frame" not the channel. I would name them as:

```
enum class image_frame {
  none,
  recordio,
  tfrecord // in the future
  ...
}
```

cbalioglu commented on include/mlio/image_reader.h:
```diff
@@ -0,0 +1,98 @@
+    /// The shape of output image in (channels, height, width) format. Region
+    /// will be cropped (centre) from the original image. Defaults to
+    /// {3, 255, 255}.
+    std::vector<int> data_shape{3, 255, 255};
```
Why not `image_shape`? As I mentioned below, I think explicitly requesting the image dimensions is a flawed logic, but even if we keep this logic temporarily, I would make this a required parameter, instead of giving some dummy values.

cbalioglu replied:
Also the dimensions should be of type `std::size_t`.

rizwangilani replied:
for (1), again trying to keep it consistent with mxnet API params: https://mxnet.apache.org/api/python/docs/api/mxnet/io/index.html#mxnet.io.ImageRecordIter

What is your opinion around having some sort of overall naming consistency with other frameworks? Do you still recommend naming it `image_shape`?

cbalioglu replied:
I would even name it `image_dimensions`. `data_shape` is too ambiguous. We should not make our API less intuitive for the sake of being consistent with MXNet.

rizwangilani replied:
makes sense

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,224 @@
+        feature_desc_builder{"value",
+                             data_type::uint8,
+                             {params().batch_size,
+                              static_cast<std::size_t>(params_.data_shape[1]),
```
Please make `data_shape`'s element type `std::size_t` instead of casting here.

rizwangilani replied:
So what I noticed was that making `data_shape` of type `std::size_t` lead to more typecasting to int at various places since opencv `Mat` used `int` for matrix rows/cols. This way, I was able to achieve cleaner code as number of typecasting decreased?

thoughts?

rizwangilani replied:
I was referring to this as the earlier comment.

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,224 @@
+{
+    auto tsr = make_tensor(batch.instances(), batch.size());
+    std::vector<intrusive_ptr<tensor>> tensors{};
+    tensors.emplace_back(tsr);
```
This should be `std::move(tsr)`. You are unnecessarily copying the tensor pointer here.

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,224 @@
+    auto instance_offset =
+        params_.data_shape[0] * params_.data_shape[1] * params_.data_shape[2];
+    auto total_memory_buffer =
+        static_cast<unsigned long>(instance_offset) * batch_size;
```
Again this should be `std::size_t`.

rizwangilani replied:
awaiting your comment on my response to your previous comment.

cbalioglu replied:
Which other comment do you mean?

rizwangilani replied:
"So what I noticed was that making data_shape of type std::size_t lead to more typecasting to int at various places since opencv Mat used int for matrix rows/cols. This way, I was able to achieve cleaner code as number of typecasting decreased?

thoughts?"

cbalioglu replied:
I would prioritize our API consistency rather than OpenCV's. The right data type to use is `std::size_t` in such cases. In our API we consistently use `std::size_t` to represent machine-indexable size entities.

As the dimension can be at most three; if you think that it causes to much casting, you can have a private instance of type `std::array<int, 3>`, copy the `image_dimensions` elements to that vector in your constructor, and use them in the rest of the code. You would have best of both worlds.

rizwangilani replied:
sounds good. done.

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,224 @@
+    auto *ptr_img_vec = img_vec.data();
+    int type;
+    cv::ImreadModes mode;
+    for (const instance &ins : instances) {
```
Our convention is `instance const &ins`. `const` should be right-handed.

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,224 @@
+    for (const instance &ins : instances) {
+        memory_slice img_buf;
+        if (params_.img_frame == image_frame::pipe) {
+            img_buf = ins.bits().subslice(recordio_image_header_offset_);
```
Not sure about this line. `recordio_record_reader` already gives you the records after shaving the headers. Why do you need this here?

rizwangilani replied:
So I spent some time on this, as I was unable to read images initially. The following is based on my understanding, and please correct me if I am wrong:

As mentioned here: https://mxnet.apache.org/api/architecture/note_data_loading - the record is divided into `kMagic`, `lrecord`, `Data`, and `pad` section. However, if you see the jpeg image example, there is a header within the `Data` section as well. This is a 24 byte header according to this: https://github.com/apache/incubator-mxnet/blob/master/src/io/image_recordio.h#L37

So I was creating the offset accordingly to skip the first 24 bytes. Please let me know what you think.

cbalioglu replied:
Oh I didn't know about that extra header. All is clear now. Ok, if we do not need to process that header, this looks fine.

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,224 @@
+    cv::Mat decoded_img = cv::imdecode(buf, mode);
+    if (decoded_img.empty()) {
+        throw std::runtime_error{fmt::format("image decoding failed. Image: ",
+                                             ins.get_data_store().id())};
```
This error message makes only sense with `file` `image_frame` option. With `pipe` (see my opinion for naming these enum values) the data store id does not represent the image. You should use a similar logic to `csv_reader` for reporting the errors with data store + record id.

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,224 @@
+                           const instance &ins) const
+{
+    cv::Mat decoded_img = cv::imdecode(buf, mode);
+    if (decoded_img.empty()) {
```
Is there no way to get more information about why decoding has failed?

rizwangilani replied:
unfortunately, `cv::imdecode()` does not return an error code. It just returns an empty matrix as documented here: https://docs.opencv.org/4.2.0/d4/da8/group__imgcodecs.html#ga26a67788faa58ade337f8d28ba0eb19e

cbalioglu commented on include/mlio/image_reader.h:
```diff
@@ -0,0 +1,98 @@
+    image_frame img_frame{image_frame::file};
+    /// Down scale the shorter edge to a new size before applying other
+    /// augmentations. Defaults to -1 to disable resize.
+    int resize = -1;
```
I assume this is a temporary option. When we infer the image dimensions from the dataset, there is no way for us to know what is the "shorter" edge. And why the shorter edge? Why not the longer edge? For resizing we should think of a more expressive API in the future.

cbalioglu replied:
I don't like sentinel values. Please use `std::optional<std::size_t>` for this parameter.

rizwangilani replied:
For (1): I was trying to keep the logic consistent with mxnet. They define resize operations has resizing the shorter edge, as mentioned here: https://mxnet.apache.org/api/python/docs/api/mxnet/io/index.html#mxnet.io.ImageRecordIter

cbalioglu replied:
Again, it is okay for now. Going forward we should come up with a better API than what MXNet provides.

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,224 @@
+        new_height = params_.resize;
+        new_width = params_.resize * src.cols / src.rows;
+    }
+    cv::resize(src, dst, cv::Size(new_width, new_height), 0, 0);
```
This operation can clearly fail. Its error status needs to be checked.

rizwangilani replied:
this doesnt seem to return any error status according to the documentation here: https://docs.opencv.org/4.2.0/da/d54/group__imgproc__transform.html#ga47a974309e9102f5f08231edc7e7529d

I have also looked at reference examples (from the documentation):
- https://docs.opencv.org/4.2.0/dd/d93/samples_2cpp_2image_alignment_8cpp-example.html#a26
- https://docs.opencv.org/4.2.0/d0/df8/samples_2cpp_2train_HOG_8cpp-example.html#a55

Neither of them are show error code usage. Thoughts?

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,224 @@
+        }
+        cv::Mat tmp = decode_image(mat, mode, ins);
+        if (params_.resize != -1) {
+            resize(tmp, tmp);
```
Is it safe to use the same memory region as both source and destination of the resize operation?

rizwangilani replied:
So usage looks safe as internally it is saving `src` and re-allocating `dst` based on the code here: https://github.com/opencv/opencv/blob/master/modules/imgproc/src/resize.cpp#L3922

This is how they seem to be using in the referenced examples as well [search for `resize`]:
- https://docs.opencv.org/4.2.0/d0/df8/samples_2cpp_2train_HOG_8cpp-example.html#a55
- https://docs.opencv.org/4.2.0/d4/db9/samples_2dnn_2object_detection_8cpp-example.html#a37 

However, if you think that relying on the method's internal implementation for memory management is bad practice then please do let me know - i'll refactor.

cbalioglu replied:
If they explicitly documented it as a safe usage, I am okay with that. However I am quite sure that this `resize` function can fail. As we discussed offline, it looks like OpenCV uses exceptions for reporting errors. I would definitely make sure that we handle them here.

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,224 @@
+            resize(tmp, tmp);
+        }
+        if (params_.to_rgb && tmp.channels() != 1) {
+            cv::cvtColor(tmp, tmp, cv::COLOR_BGR2RGB);
```
I assume this operation can potentially fail. You should check its error status.

rizwangilani replied:
method doesn't return an error code: https://docs.opencv.org/4.2.0/db/d64/tutorial_js_colorspaces.html

cbalioglu replied:
As discussed offline, OpenCV uses exceptions for error propagation.

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,224 @@
+}
+
+void
+image_reader::crop(cv::Mat &src, cv::Mat &dst) const
```
Again, I assume this is a temporary mitigation. As an API caller, I wouldn't want my images to get cropped.

rizwangilani replied:
This is again taken from the logic from mxnet. Mxnet adds additional adds the option for random crop as well. The only way to create images of uniform size is to:

1. crop them
2. resize them even if that is against the aspect ratio

I was going with the option (1)

cbalioglu replied:
Sounds good.

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,224 @@
+    int x = (src.cols - params_.data_shape[2]) / 2;
+
+    cv::Rect roi(x, y, params_.data_shape[2], params_.data_shape[1]);
+    src(roi).copyTo(dst);
```
Like other places, this operation can fail, and we do not check it.

rizwangilani replied:
this method does not return any error code: https://docs.opencv.org/4.2.0/d3/d63/classcv_1_1Mat.html#a33fd5d125b4c302b0c9aa86980791a77

cbalioglu reviewed (commented):
Two points besides inline feedback.

1. Right now ML-IO fails to build if I don't enable the `MLIO_BUILD_IMAGE_READER` CMake flag.
2. As we discussed offline, I think this code should reside in the contrib package for the time-being as it is in a very early stage and lacks many essential features to be a generic image reader.

rizwangilani commented:
> Two points besides inline feedback.
> 
> 1. Right now ML-IO fails to build if I don't enable the `MLIO_BUILD_IMAGE_READER` CMake flag.
> 2. As we discussed offline, I think this code should reside in the contrib package for the time-being as it is in a very early stage and lacks many essential features to be a generic image reader.

for (2), how about I do that in a separate commit? it would result in a cleaner PR in my opinion, and would allow us to focus on core functionality for this commit.

rizwangilani commented:
1. I have updated the PR with the latest code after addressing the resolved comments - incase you have additional feedback.

2. For the purpose of completion/testing, I have also add changes related to https://github.com/rizwangilani/ml-io/pull/2 in this PR.

3. Will push another update once we resolve the remaining (open) comments.

cbalioglu commented on include/mlio/image_reader.h:
```diff
@@ -0,0 +1,100 @@
+private:
+    intrusive_ptr<schema> schema_{};
+    image_reader_params params_{};
+    size_t static const image_dimensions_size_ = 3;
```
Please use:
```
static std::size_t constexpr image_dimensions_size_ = 3;
static std::size_t constexpr recordio_image_header_offset_ = 24;
```

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,316 @@
+{
+    if (params_.image_dimensions.size() != image_dimensions_size_) {
+        throw std::invalid_argument{
+            "image_dimensions is a required argument. "
```
Minor fix in wording: "image_dimensions is a required parameter". Parameter is what you see in the function signature. Argument is the actual value that you pass to a parameter.

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,316 @@
+#ifdef MLIO_BUILD_IMAGE_READER
+
+#include <fmt/format.h>
+#include <opencv4/opencv2/imgcodecs.hpp>
```
"opencv4" part of the path fails for me when I try to build MLIO in a Conda environment. Are you sure that this is the right include path that is used by CMake internally. You can open CMakeLists.txt file and see where it points for OpenCV.

rizwangilani replied:
I rechecked and it seems to be working for me. let's connect offline to resolve this.

cbalioglu commented on include/mlio/image_reader.h:
```diff
@@ -0,0 +1,100 @@
+    /// The dimensions of output image in (channels, height, width) format.
+    /// Region will be cropped (centre) from the original image. Defaults to
+    /// {3, 255, 255}.
+    std::vector<int> image_dimensions{3, 255, 255};
```
Again, do we really need these dummy dimensions if this parameter is now required and validated in the constructor?

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,316 @@
+#include "mlio/record_readers/blob_record_reader.h"
+#include "mlio/record_readers/recordio_record_reader.h"
+#include "mlio/schema.h"
+#include "mlio/streams/utf8_input_stream.h"
```
Is this include used somewhere?

rizwangilani replied:
removed. It was being used before. thanks for pointing it out.

cbalioglu commented on tests/mlio-test/test_image_reader.cxx:
```diff
@@ -0,0 +1,445 @@
+#include <gtest/gtest.h>
+#include <mlio.h>
+
+#ifdef MLIO_BUILD_IMAGE_READER
```
This file is only used if `MLIO_BUILD_IMAGE_READER` is defined. Instead of using `#ifdef`, I recommend managing this in CMake. Here is an example:
https://github.com/awslabs/ml-io/blob/master/src/mlio/CMakeLists.txt#L83

rizwangilani replied:
the quoted line 83 has `shuffled_instance_reader.cxx` - is that what you meant to refer to?

cbalioglu reviewed (commented):
One thing that I forgot to mention in my first review is the handling of `bad_batch_handling` and `last_batch_handling` parameters. Similar to `csv_reader` and `recordio_protobuf_reader`, you should respect those parameters in your implementation.

rizwangilani commented:
Addressed comments. Added `bad_batch_handling`. `last_batch_handling` logic was already there. I am not sure why `opencv4` path is not working. Let's sync up offline.

Please let me know if you have any more comments.

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,392 @@
+image_reader::make_tensor(std::vector<instance> const &instances,
+                          std::size_t batch_size) const
+{
+    std::atomic_bool skip_batch{};
```
This can be a regular `bool`. You don't have any multithreading here, so you don't need an atomic. There is by the way a low hanging fruit here that can significantly boost the performance of the reader. Right now you are decoding images sequentially. You can easily parallelize those decoding operations. See other readers to see how you can use Intel TBB for that. Not necessary for this PR, but you should consider it in a feature update.

rizwangilani replied:
got it, thanks for the suggestion.

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,392 @@
+    for (instance const &ins : instances) {
+        memory_slice img_buf;
+        if (params_.img_frame == image_frame::recordio) {
+            img_buf = ins.bits().subslice(recordio_image_header_offset_);
```
Could you please write a short comment here why we are skipping the header? I am sure in not so distant feature someone will ask the exact same question that I asked in my first review.

cbalioglu commented on include/mlio/image_reader.h:
```diff
@@ -0,0 +1,103 @@
+    image_frame img_frame{image_frame::none};
+    /// Down scale the shorter edge to a new size before applying other
+    /// augmentations. Defaults to -1 to disable resize.
+    int resize = -1;
```
I still prefer to have this as `std::optional<std::size_t> resize{};` instead of using a magic number and to be consistent with other `*_params` classes.

rizwangilani replied:
addressed. You think it should be an optional parameter? Isn't this something that user should be aware of, especially if we are removing `no resize` as default?

rizwangilani replied:
done, made it optional.

cbalioglu commented on include/mlio/image_reader.h:
```diff
@@ -0,0 +1,103 @@
+    bad_batch_handling bbh_{};
+    intrusive_ptr<schema> schema_{};
+    image_reader_params params_{};
+    static std::size_t constexpr image_dimensions_size_ = 3;
```
Minor: I usually put `static constexpr` statements to the top of the class declaration for better visibility and use the bottom part only for instance variable declarations. For an example see [here](https://github.com/awslabs/ml-io/blob/master/include/mlio/record_readers/parquet_record_reader.h#L41).

rizwangilani replied:
that makes a lot of sense

cbalioglu commented on include/mlio/image_reader.h:
```diff
@@ -0,0 +1,103 @@
+    operator=(image_reader &&) = delete;
+
+private:
+    intrusive_ptr<record_reader>
```
Minor: Please put an `MLIO_HIDDEN` to the private function declarations. See [here](https://github.com/awslabs/ml-io/blob/master/include/mlio/recordio_protobuf_reader.h#L62) for an example. This avoid unnecessarily exporting the symbols for private functions in the shared library.

cbalioglu commented on include/mlio/image_reader.h:
```diff
@@ -0,0 +1,103 @@
+class MLIO_API image_reader final : public parallel_data_reader {
+public:
+    explicit image_reader(data_reader_params rdr_prm,
+                          image_reader_params image_prm = {});
```
`image_prm` should not have a default value as it contains required parameters. The user *must* pass it to the constructor in all cases.

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,392 @@
+    case image_frame::recordio:
+        return make_intrusive<detail::recordio_record_reader>(ds.open_read());
+    }
+    throw std::invalid_argument{"Please enter a valid value for "
```
Minor: In the rest of the code base the error messages are written in a way that describe the error state rather than referring to the user. Here for example it would be more consistent to have an error message like: `The img_frame argument does not represent a valid value.`.

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,392 @@
+            break;
+        default:
+            throw std::invalid_argument{
+                fmt::format("Unsupported number of channels: {0:n}",
```
Make sure to include the `data_store` and `index` information for better diagnosability. Also please try to write whole sentences to be consistent with other error message. For example:
`The record X in the data store Y has Z channels while the maximum number of supported channels is 3.`

rizwangilani replied:
So this will be a user error. User would have passed an invalid argument for the `image_dimensions_` parameter. That is why it wouldnt make sense for the error message to talk about the `data_store` or `index`.

rizwangilani replied:
Opinion?

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,392 @@
+                            params_.image_dimensions[0])};
+        }
+        cv::Mat tmp = decode_image(mat, mode, ins, skip_batch);
+        if (skip_batch) {
```
Instead of passing and checking `skip_batch`, can't we check whether `tmp` is `empty` and consider it as a failed decoding?

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,392 @@
+            return nullptr;
+        }
+        if (params_.resize != -1) {
+            resize(tmp, tmp, skip_batch);
```
Similarly, instead of passing `skip_batch`, I would have a return value of `bool` for the `resize` function.

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,392 @@
+                    params_.image_dimensions[2],
+                    type,
+                    ptr_img_vec};
+        crop(tmp, dst, skip_batch);
```
Same here. Returning `bool` from `crop` would be cleaner.

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,392 @@
+                           image_reader_params image_prm)
+    : parallel_data_reader{std::move(rdr_prm)}, params_{std::move(image_prm)}
+{
+    bbh_ = {};
```
Move this to the constructor initializer list, next to `params_` in line 325.

rizwangilani replied:
Moving it there gives a warning saying: `warning: private field 'img_dims_' is not used [clang-diagnostic-unused-private-field]`

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,392 @@
+intrusive_ptr<record_reader>
+image_reader::make_record_reader(data_store const &)
+{
+   throw not_supported_error{"ML-IO has not been built with image reader "
```
For all member functions except the constructor you don't need the `not_supported_error` statement as it is impossible to construct an `image_reader` instance anyway. Instead of throwing an error, simply do nothing if the return is `void`, or return a dummy value otherwise (e.g. `return nullptr;`)

cbalioglu reviewed (commented):
Mostly a couple minor comments. We are almost there. Besides my inline comments two other things to mention:

1. Especially test_image_0.png is quite large to commit to the repo (125KB). Is it possible to use a smaller PNG file? The other image files are relatively small, but I wouldn't say no, if you could find smaller versions of them as well.

2. I built the code with clang-tidy enabled and got the following output. Mostly minor things, but please address them in your next revision as well.

```
/home/ubuntu/repos/ml-io/build/release-clang/../../src/mlio/image_reader.cxx:108:21: warning: consider replacing 'unsigned long' with 'uint64' [google-runtime-int]
        static_cast<unsigned long>(instance_offset) * batch_size;
                    ^
/home/ubuntu/repos/ml-io/build/release-clang/../../src/mlio/image_reader.cxx:111:9: warning: variable 'type' is not initialized [cppcoreguidelines-init-variables]
    int type;
        ^
             = 0
/home/ubuntu/repos/ml-io/build/release-clang/../../src/mlio/image_reader.cxx:112:21: warning: variable 'mode' is not initialized [cppcoreguidelines-init-variables]
    cv::ImreadModes mode;
                    ^
                         = 0
/home/ubuntu/repos/ml-io/build/release-clang/../../src/mlio/image_reader.cxx:125:33: warning: do not use const_cast [cppcoreguidelines-pro-type-const-cast]
            static_cast<void *>(const_cast<std::byte *>(img_buf.data())));
                                ^
/home/ubuntu/repos/ml-io/build/release-clang/../../src/mlio/image_reader.cxx:235:5: warning: multiple declarations in a single statement reduces readability [readability-isolate-declaration]
    int new_height, new_width;
    ^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/ubuntu/repos/ml-io/build/release-clang/../../src/mlio/image_reader.cxx:235:9: warning: variable 'new_height' is not initialized [cppcoreguidelines-init-variables]
    int new_height, new_width;
        ^
                   = 0
/home/ubuntu/repos/ml-io/build/release-clang/../../src/mlio/image_reader.cxx:235:21: warning: variable 'new_width' is not initialized [cppcoreguidelines-init-variables]
    int new_height, new_width;
                    ^
                              = 0
[2/5] Building CXX object tests/mlio-test/CMakeFiles/mlio-test.dir/test_image_reader.cxx.o
/home/ubuntu/repos/ml-io/build/release-clang/../../include/mlio/span.h:202:13: warning: use default member initializer for 'data_' [modernize-use-default-member-init]
    pointer data_;
            ^
                 {}
/home/ubuntu/repos/ml-io/build/release-clang/../../tests/mlio-test/test_image_reader.cxx:59:13: warning: multiple declarations in a single statement reduces readability [readability-isolate-declaration]
            int new_height, new_width;
            ^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/ubuntu/repos/ml-io/build/release-clang/../../tests/mlio-test/test_image_reader.cxx:59:17: warning: variable 'new_height' is not initialized [cppcoreguidelines-init-variables]
            int new_height, new_width;
                ^
                           = 0
/home/ubuntu/repos/ml-io/build/release-clang/../../tests/mlio-test/test_image_reader.cxx:59:29: warning: variable 'new_width' is not initialized [cppcoreguidelines-init-variables]
            int new_height, new_width;
                            ^
                                      = 0
/home/ubuntu/repos/ml-io/build/release-clang/../../tests/mlio-test/test_image_reader.cxx:78:5: warning: method 'assert_schema' can be made const [readability-make-member-function-const]
    assert_schema(const schema &sch, size_t batch_size, int channel)
    ^
                                                                     const
/home/ubuntu/repos/ml-io/build/release-clang/../../tests/mlio-test/test_image_reader.cxx:87:5: warning: method 'assert_tensor_shape' can be made const [readability-make-member-function-const]
    assert_tensor_shape(dense_tensor *tsr, size_t batch_size, int channel)
    ^
                                                                           const
/home/ubuntu/repos/ml-io/build/release-clang/../../tests/mlio-test/test_image_reader.cxx:365:12: warning: catch handler catches by value; should catch by reference instead [cert-err09-cpp]
    catch (std::invalid_argument ex) {
           ^
/home/ubuntu/repos/ml-io/build/release-clang/../../tests/mlio-test/test_image_reader.cxx:365:12: warning: catch handler catches by value; should catch by reference instead [cert-err61-cpp]
/home/ubuntu/repos/ml-io/build/release-clang/../../tests/mlio-test/test_image_reader.cxx:365:12: warning: catch handler catches by value; should catch by reference instead [misc-throw-by-value-catch-by-reference]
/home/ubuntu/repos/ml-io/build/release-clang/../../tests/mlio-test/test_image_reader.cxx:389:5: warning: multiple declarations in a single statement reduces readability [readability-isolate-declaration]
    cv::Mat expected_image{get_expected_image(img0_path_png_, resize)}, output;
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[3/5] Linking CXX shared library lib/libmlio.so.0.3.2
[4/5] Creating library symlink lib/libmlio.so.0 lib/libmlio.so
[5/5] Linking CXX executable tests/mlio-test/mlio-test
```

rizwangilani commented:
Addressed all comments. Please let me know if you have more. thanks a lot.

Updated python bindings too: https://github.com/rizwangilani/ml-io/pull/4

cbalioglu commented on include/mlio/image_reader.h:
```diff
@@ -0,0 +1,112 @@
+    /// Optional parameter to down scale the shorter edge to a new size
+    /// before applying other augmentations.
+    std::optional<std::size_t> resize{};
+    /// The dimensions of output image in (channels, height, width) format.
```
Minor: Please limit the comment line lengths to 72.

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,393 @@
+            try {
+                cv::cvtColor(tmp, tmp, cv::COLOR_BGR2RGB);
+            }
+            catch (cv::Exception &e) {
```
Minor: You should catch exceptions by const reference.
```
catch (cv::Exception const &e) { .. }
```

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,393 @@
+                           int mode,
+                           instance const &ins) const
+{
+    cv::Mat decoded_img = cv::imdecode(buf, mode);
```
Shouldn't this call be wrapped in a try/catch block?

rizwangilani replied:
I initially skipped that is documentation suggested that it will return an empty mat incase of invalid data or buffer (tested it as well). However, it may be safer to wrap it within try catch. updated.

cbalioglu reviewed (commented):
Just three minor comments. Let's sync up online about the OpenCV header file path issue.

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,411 @@
+}
+
+void
+image_reader::infer_schema(instance const &)
```
As I mentioned offline, there have been two minor changes in the API:
- `infer_schema` now expects to return the `schema` object and it accepts the `instance` as an optional parameter.
- `feature_desc` is renamed to `attribute`.

Now with these changes this implementation should look like the following:

```c++
std::intrusive_ptr<schema>
image_reader::infer_schema(std::optional<instance> const &)
{
  std::vector<attribute> attrs{};
  attrs.emplace_back("value", data_type::uint8, size_vector{...});
  return make_intrusive<schema>(std::move(attrs));
}
```

rizwangilani replied:
done

cbalioglu commented on include/mlio/image_reader.h:
```diff
@@ -0,0 +1,110 @@
+
+private:
+    bad_batch_handling bbh_{};
+    intrusive_ptr<schema> schema_{};
```
See my feedback on `infer_schema`. This variable can now be removed with the latest API changes.

rizwangilani replied:
done

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,411 @@
+    }
+    std::vector<intrusive_ptr<tensor>> tensors{};
+    tensors.emplace_back(std::move(tsr));
+    return make_intrusive<example>(schema_, std::move(tensors));
```
Instead of `schema_` use the protected `get_schema` member function of `parallel_data_reader` to access the schema.

rizwangilani replied:
done

cbalioglu reviewed (commented):
Looks good to me. I left three inline comments that explain the minor API changes in the latest code base. Once you rebase your PR to the head of the master branch, you should make those small changes.

As the next step I expect you to send your "real" PR to the main repo.

Commit:
Change data store instance indexing to start from 0

rizwangilani commented:
- Addressed all comments.
- rebased with mlio refactoring
- updated python bindings and fixed unit tests (sep commits): https://github.com/rizwangilani/ml-io/pull/4

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,413 @@
+{
+    std::vector<attribute> attrs{};
+    // schema follows the NHWC convention
+    attrs.emplace_back(attribute_builder{"value",
```
Minor. This can be reduced to:
```
    attrs.emplace_back("value",
                                       data_type::uint8,
                                       size_vector{params().batch_size,
                                       params_.image_dimensions[1],
                                       params_.image_dimensions[2],
                                       params_.image_dimensions[0]});
```

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,413 @@
+
+image_reader::image_reader(data_reader_params rdr_prm,
+                           image_reader_params image_prm)
+    : parallel_data_reader{std::move(rdr_prm)}, params_{std::move(image_prm)}
```
Minor. You can put all member variable initializations to the list.
```
: parallel_data_reader{std::move(rdr_prm)}, params_{std::move(image_prm)}, bbh_{}, img_dims_{}
{
...
}
```

rizwangilani replied:
resolved as discussed offline.

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,413 @@
+                           instance const &) const
+{
+    cv::Mat foo;
+    return foo;
```
Minor. `return {};` would work.

cbalioglu commented on src/mlio/image_reader.cxx:
```diff
@@ -0,0 +1,413 @@
+}
+
+intrusive_ptr<dense_tensor>
+image_reader::make_tensor(std::vector<instance> const &,
```
Minor. Both `make_tensor` and `decode_image` can fit into a single line as they don't have argument names.
```
image_reader::make_tensor(std::vector<instance> const &, std::size_t) const
{
...
}
```

cbalioglu reviewed (approved):
Just have a couple very minor comments. Otherwise looks good to me.
